File: FirstFitMemoryAllocation.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class Frame {
    int frameSize;
    boolean isOccupied;
    int pageId;

    Frame(int frameSize) {
        this.frameSize = frameSize;
        this.isOccupied = false;
        this.pageId = -1;
    }
}

public class FirstFitMemoryAllocation {
    private final List<Frame> frames;

    public FirstFitMemoryAllocation(List<Integer> frameSizes) {
        frames = new ArrayList<>();
        // Initialize frames with user-defined sizes
        for (int frameSize : frameSizes) {
            frames.add(new Frame(frameSize));
        }
    }

    public boolean allocateMemory(int pageId, int pageSize) {
        // Check if the page is already allocated
        for (Frame frame : frames) {
            if (frame.isOccupied && frame.pageId == pageId) {
                System.out.println("Page " + pageId + " is already allocated.");
                return false;
            }
        }

        // Find the first suitable frame for the page
        for (Frame frame : frames) {
            if (!frame.isOccupied && frame.frameSize >= pageSize) {
                frame.isOccupied = true;
                frame.pageId = pageId;
                System.out.println("Allocated page " + pageId + " of size " + pageSize + " into frame of size " + frame.frameSize);
                return true;
            }
        }

        System.out.println("No suitable frame found for page " + pageId + " with size " + pageSize);
        return false;
    }

    public void freeMemory(int pageId) {
        for (Frame frame : frames) {
            if (frame.isOccupied && frame.pageId == pageId) {
                frame.isOccupied = false;
                frame.pageId = -1;
                System.out.println("Freed memory for page " + pageId);
                return;
            }
        }
        System.out.println("Page " + pageId + " is not found in memory.");
    }

    public void displayMemoryStatus() {
        System.out.println("\nMemory Status:");
        for (int i = 0; i < frames.size(); i++) {
            Frame frame = frames.get(i);
            String status = frame.isOccupied ? "Occupied by page " + frame.pageId : "Free";
            System.out.println("Frame " + i + " (Size: " + frame.frameSize + "): " + status);
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Get memory block sizes from the user
        List<Integer> frameSizes = new ArrayList<>();
        System.out.print("Enter the number of memory blocks: ");
        int numBlocks = scanner.nextInt();
        
        System.out.println("Enter the sizes of each memory block:");
        for (int i = 0; i < numBlocks; i++) {
            System.out.print("Size of block " + (i + 1) + ": ");
            frameSizes.add(scanner.nextInt());
        }

        FirstFitMemoryAllocation memory = new FirstFitMemoryAllocation(frameSizes);

        // User menu for memory operations
        boolean exit = false;
        while (!exit) {
            System.out.println("\nChoose an operation:");
            System.out.println("1. Allocate Memory");
            System.out.println("2. Free Memory");
            System.out.println("3. Display Memory Status");
            System.out.println("4. Exit");
            System.out.print("Enter your choice: ");
            int choice = scanner.nextInt();

            switch (choice) {
                case 1:
                    System.out.print("Enter the page ID to allocate: ");
                    int pageId = scanner.nextInt();
                    System.out.print("Enter the size of the page: ");
                    int pageSize = scanner.nextInt();
                    memory.allocateMemory(pageId, pageSize);
                    break;

                case 2:
                    System.out.print("Enter the page ID to free: ");
                    int freePageId = scanner.nextInt();
                    memory.freeMemory(freePageId);
                    break;

                case 3:
                    memory.displayMemoryStatus();
                    break;

                case 4:
                    exit = true;
                    System.out.println("Exiting...");
                    break;

                default:
                    System.out.println("Invalid choice. Please try again.");
            }
        }

        scanner.close();
    }
}

File: FirstFitMemoryAllocation.py
class Frame:
    def __init__(self, frame_size):
        self.frame_size = frame_size
        self.is_occupied = False
        self.page_id = None

class FirstFitMemoryAllocation:
    def __init__(self, frame_sizes):
        # Initialize frames with user-defined sizes
        self.frames = [Frame(size) for size in frame_sizes]

    def allocate_memory(self, page_id, page_size):
        # Check if page is already loaded
        for frame in self.frames:
            if frame.is_occupied and frame.page_id == page_id:
                print(f"Page {page_id} is already allocated.")
                return False

        # Find the first suitable frame for the page
        for frame in self.frames:
            if not frame.is_occupied and frame.frame_size >= page_size:
                frame.is_occupied = True
                frame.page_id = page_id
                print(f"Allocated page {page_id} of size {page_size} into frame of size {frame.frame_size}")
                return True

        print(f"No suitable frame found for page {page_id} with size {page_size}")
        return False

    def free_memory(self, page_id):
        # Free the frame occupied by the page with the given page_id
        for frame in self.frames:
            if frame.is_occupied and frame.page_id == page_id:
                frame.is_occupied = False
                frame.page_id = None
                print(f"Freed memory for page {page_id}")
                return
        print(f"Page {page_id} is not found in memory.")

    def display_memory_status(self):
        print("\nMemory Status:")
        for i, frame in enumerate(self.frames):
            status = f"Occupied by page {frame.page_id}" if frame.is_occupied else "Free"
            print(f"Frame {i} (Size: {frame.frame_size}): {status}")

def main():
    # Get memory block sizes from the user
    frame_sizes = []
    num_blocks = int(input("Enter the number of memory blocks: "))
    
    print("Enter the sizes of each memory block:")
    for i in range(num_blocks):
        block_size = int(input(f"Size of block {i + 1}: "))
        frame_sizes.append(block_size)

    memory = FirstFitMemoryAllocation(frame_sizes)

    # User menu for memory operations
    while True:
        print("\nChoose an operation:")
        print("1. Allocate Memory")
        print("2. Free Memory")
        print("3. Display Memory Status")
        print("4. Exit")
        choice = int(input("Enter your choice: "))

        if choice == 1:
            page_id = int(input("Enter the page ID to allocate: "))
            page_size = int(input("Enter the size of the page: "))
            memory.allocate_memory(page_id, page_size)

        elif choice == 2:
            page_id = int(input("Enter the page ID to free: "))
            memory.free_memory(page_id)

        elif choice == 3:
            memory.display_memory_status()

        elif choice == 4:
            print("Exiting...")
            break

        else:
            print("Invalid choice. Please try again.")

if __name__ == "__main__":
    main()
